### PR TITLE
feat(api): add `nulls_first=False` argument to `order_by`

### PR DESCRIPTION
**Pull request by ncclementi**
- Closes https://github.com/ibis-project/ibis/issues/7252

This is work in progress

## TODO

- [x] Sort out (pun intended) how to deal with backends that don't expose option for ordering nulls. 
        - [x]   mssql
        - [x]   mysql 
EDIT: They get handled by sqlglot and follow the expected behavior

- [x] Add test in `test_api.py` to check that this `t.order_by(ibis.desc("b", nulls_first=True))` is equal than `t.order_by(t["b"].desc(nulls_first=True))`
- [x] fix window tests errors (Thanks @gforsyth for your the "diving into deep into the weeds" session)
 
- [x] Decide how and deprecate https://github.com/ibis-project/ibis/blob/2dac5e48d80e4251d3d682307ff90e00e52455f5/ibis/expr/operations/sortkeys.py#L38-L51

Options are: 
- deprecate passing a tuple completely
- support 2-tuples and 3-tuples (no deprecation)
- don't support 3-tuples, leave 2-tuples for legacy reasons. 
 
NOTE: I can't find any tests using this, not clear to me what do we want to do here. 
EDIT: see https://github.com/ibis-project/ibis/issues/9413

- [x] Figure out why the doctests and plenty others are failing,I think it's because if we don't say `desc()` or `asc()` is taking the `nulls_first` default value set in `SortKeys` which is `True` (I'm not sure yet, need to investigate) **EDIT:** doctest and other tests using order_by that didn't trigger `desc` or `asc` rely on defaul `nulls_first` in `SortKeys`. Before this value wasn't provided and SQLGlot handel this by triggering nulls_first=False. So to be consistent we set it to False in the class default. 

~~I know why they are failing, it's unclear what path we want to take here. Need to discuss~~

https://github.com/ncclementi/ibis/blob/78f26dada233a3da3a5e145fc32d590e058f1a4e/ibis/expr/operations/sortkeys.py#L28-L37

- [x] Discuss/fix: Now we have a lot of snapshot tests failing, because of the fix in the `window_merge_frames` rewrite, where we are now passing `nulls_first` see 
https://github.com/ibis-project/ibis/blob/a17ff2ec4eae0629fbfc00294b51c2089be8faf1/ibis/expr/rewrites.py#L310-L314

- [x] Add tests that orders by more than one column, has nulls, and with different value of nulls_first.

### CODE REVIEW

**Comment by cpcloud** [comment]
Fixing up the snapshots now.

File: ibis/backends/pandas/executor.py [hunk 1]
```diff
@@ -593,11 +593,18 @@ def visit(cls, op: ops.Sort, parent, keys):
         # 2. sort the dataframe using those columns
         # 3. drop the sort key columns
         ascending = [key.ascending for key in op.keys]
+        na_pos_dict = {True: "first", False: "last"}
```
Comment (ncclementi) [inline ibis/backends/pandas/executor.py:596]: I'm not sure this is the best way to pass `na_position`. Same hack than in the dask executor.

File: ibis/backends/polars/compiler.py [hunk 2]
```diff
@@ -227,10 +227,14 @@ def sort(op, **kw):
 
     by = list(newcols.keys())
     descending = [key.descending for key in op.keys]
+    # NOT SURE IF THIS IS THE RIGHT THING TO DO
+    nulls_first = all(key.nulls_first for key in op.keys)
```
Comment (ncclementi) [inline ibis/backends/polars/compiler.py:231]: Not sure if this is the best way to go about it.

File: ibis/backends/polars/compiler.py [hunk 3]
```diff
@@ -227,10 +227,14 @@ def sort(op, **kw):
...
     try:
-        lf = lf.sort(by, descending=descending, nulls_last=True)
+        lf = lf.sort(by, descending=descending, nulls_last=not nulls_first)
     except TypeError:  # pragma: no cover
-        lf = lf.sort(by, reverse=descending, nulls_last=True)  # pragma: no cover
+        lf = lf.sort(
```
Comment (ncclementi) [inline ibis/backends/polars/compiler.py:235]: While testing this, I noticed that the `reverse` keyword doesn't work here. I'm not sure what are we trying to do, this is doesn't seem to be tested so I'm not sure what behavior we want.

File: ibis/backends/dask/executor.py [hunk 4]
```diff
@@ -364,10 +364,17 @@ def visit(cls, op: ops.Sort, parent, keys):
         # 2. sort the dataframe using those columns
         # 3. drop the sort key columns
         ascending = [key.ascending for key in op.keys]
+        na_pos_dict = {True: "first", False: "last"}
+        na_position = na_pos_dict[all(key.nulls_first for key in op.keys)]
```
Comment (ncclementi) [inline ibis/backends/dask/executor.py:368]: pandas doesn't support a list that contains different options for null orders. It only accepts `na_position`= "last" or "first".

What do we  want to do here? we could check that if there is more than one to break.

File: ibis/backends/sql/dialects.py [hunk 5]
```diff
@@ -336,6 +338,7 @@ def _create_sql(self, expression: sge.Create) -> str:
     return self.create_sql(expression)
 
 
+Oracle.NULL_ORDERING = "nulls_are_large"
```
Comment (ncclementi) [inline ibis/backends/sql/dialects.py:341]: I opened the issue in sqlglot but I'm not sure if I explained it right. Feel free to chime in
https://github.com/tobymao/sqlglot/issues/3684

File: ibis/backends/sql/dialects.py [hunk 6]
```diff
@@ -336,6 +338,7 @@ def _create_sql(self, expression: sge.Create) -> str:
     return self.create_sql(expression)
```
Comment (ncclementi) [inline ibis/backends/sql/dialects.py:340]:
```suggestion
Hack around https://github.com/tobymao/sqlglot/issues/3684
```

File: ibis/backends/sql/dialects.py [hunk 7]
```diff
@@ -335,7 +337,8 @@ def _create_sql(self, expression: sge.Create) -> str:
             )
     return self.create_sql(expression)
 
-
+Hack around https://github.com/tobymao/sqlglot/issues/3684
```
Comment (ncclementi) [inline ibis/backends/sql/dialects.py:340]:
```suggestion
# Hack around https://github.com/tobymao/sqlglot/issues/3684
```

File: ibis/expr/operations/sortkeys.py [hunk 8]
```diff
@@ -40,15 +41,15 @@ def __coerce__(cls, value, T=None, S=None):
         if isinstance(value, tuple):
```
Comment (ncclementi) [inline ibis/expr/operations/sortkeys.py:41]: see: https://github.com/ibis-project/ibis/issues/9413

File: ibis/backends/dask/executor.py [hunk 9]
```diff
@@ -364,10 +364,23 @@ def visit(cls, op: ops.Sort, parent, keys):
...
         # 3. drop the sort key columns
         ascending = [key.ascending for key in op.keys]
+        nulls_first = [key.nulls_first for key in op.keys]
+
+        if all(nulls_first):
+            na_position = "first"
+        elif not any(nulls_first):
+            na_position = "last"
+        else:
+            raise ValueError("dask does not support different columns ordering")
```
Comment (ncclementi) [inline ibis/backends/dask/executor.py:374]: We currently don't have any tests checking for ordering more than one column with different values of nulls_first. My guess is that we should add this kind of test in test_generic.

**Comment by ncclementi** [comment]
Need to add tests that check for order by more than one column, has nulls, and with different value of `nulls_first`.

File: ibis/backends/dask/executor.py (hunk 9 above)
Comment (cpcloud) [reply ibis/backends/dask/executor.py:374]: Yep, that sounds right!

Comment (cpcloud) [inline ibis/backends/dask/executor.py:374]:
```suggestion
            raise ValueError("dask does not support specifying null ordering per column")
```

File: ibis/backends/pandas/executor.py [hunk 10]
```diff
@@ -593,11 +593,24 @@ def visit(cls, op: ops.Sort, parent, keys):
...
         # 3. drop the sort key columns
         ascending = [key.ascending for key in op.keys]
+        nulls_first = [key.nulls_first for key in op.keys]
+
+        if all(nulls_first):
+            na_position = "first"
+        elif not any(nulls_first):
+            na_position = "last"
+        else:
+            raise ValueError("pandas does not support different columns ordering")
```
Comment (cpcloud) [inline ibis/backends/pandas/executor.py:603]:
```suggestion
            raise ValueError("pandas does not support specifying null ordering for individual columns")
```

**Review by cpcloud — APPROVED** [review]
Great work!

We can add the tests for multiple `nulls_first` specifications here or in a follow-up, up to you.

**Comment by ncclementi** [comment]
> We can add the tests for multiple nulls_first specifications here or in a follow-up, up to you.

I'll added it here for consistency, working on it now. I'll ping you when done.

File: ibis/backends/tests/test_generic.py [hunk 11]
```diff
@@ -599,6 +600,126 @@ def test_order_by_random(alltypes):
...
+    if con.name == "pandas" and (nf1 != nf2):
+        with pytest.raises(
+            ValueError, match="pandas does not support different columns ordering"
+        ):
+            result = con.execute(expr).reset_index(drop=True)
+    else:
+        result = con.execute(expr).reset_index(drop=True)
+        expected = pd.DataFrame(expected)
+
+        tm.assert_frame_equal(result.replace({np.nan: None}), expected)
```
Comment (ncclementi) [inline ibis/backends/tests/test_generic.py:720]: If I do this locally `.replace({np.nan: None})` I get this

```python
>           tm.assert_frame_equal(result.replace({np.nan: None}), expected)
E           AssertionError: Attributes of DataFrame.iloc[:, 0] (column name="col1") are different
E           
E           Attribute "dtype" are different
E           [left]:  object
E           [right]: float64
```

only in this test. I don't see it in the other ones.

Comment (cpcloud) [reply ibis/backends/tests/test_generic.py:721]: It's the `expected` that needs the `astype`, not the `result`. I'll push up a fix.

Comment (ncclementi) [reply ibis/backends/tests/test_generic.py]: Yes, what's not clear to me why in the other tests we needed it. Locally for me, they all passed without it.

**Comment by cpcloud** [comment]
After losing a few brain cells trying to understand how to set collation and sorting in MS SQL, we gave up and decided to xfail the multi key sort column null ordering tests for backends that for some reason known only to the original developers that strings should compare ignoring case **by default**.

And it should be outrageously difficult to change that behavior.

**Pull request merged**